server/src/main/java/org/cloudfoundry/identity/uaa/client/VmidentityClientMetadataProvisioning.java

```diff
@@ -0,0 +1,144 @@
+            return clientMetadata;
+        } catch (Exception e) {
+            logger.error("Unable to retrieve UI details for all clients", e);
+            throw new VmidentityDataAccessException("Unable to retrieve UI details for all clients");
+        }
+    }
+
+    @Override
+    public ClientMetadata retrieve(String clientId) {
+        try {
+            String tenant = VmidentityUtils.getTenantName(idmClient);
+            OIDCClient client = idmClient.getOIDCClient(tenant, clientId);
+            return mapOIDCClient(client, tenant);
+        } catch (Exception e) {
+            logger.error("Unable to retrieve UI details for client: " + clientId, e);
+            throw new VmidentityDataAccessException("Unable to retrieve UI details for client: " + clientId);
+        }
+    }
+
+    @Override
+    public ClientMetadata update(ClientMetadata resource) {
+        logger.debug("Updating metadata for client: " + resource.getClientId());
+
+        try {
+            String tenant = VmidentityUtils.getTenantName(idmClient);
+            OIDCClient client = idmClient.getOIDCClient(tenant, resource.getClientId());
+
+            OIDCClient.Builder builder = new OIDCClient.Builder(client);
+
+            Map<String, Object> additionalInformation = client.getAdditionalInformation();
+            if (additionalInformation == null) {
+                additionalInformation = new HashMap<String, Object>();
+            }
+            if (resource.getClientName() != null) {
+                additionalInformation.put(ClientConstants.CLIENT_NAME, resource.getClientName());
+            }
+
+            if (resource.isShowOnHomePage()) {
+                additionalInformation.put(CLIENT_METADATA_SHOW_ON_HOME_PAGE, resource.isShowOnHomePage());
+            }
+
+            if (resource.getAppLaunchUrl() != null) {
+                additionalInformation.put(CLIENT_METADATA_APP_LAUNCH_URL, resource.getAppLaunchUrl());
+            }
+
+            if (resource.getAppIcon() != null) {
+                additionalInformation.put(CLIENT_METADATA_APP_ICON, resource.getAppIcon());
+            }
+
+            if (!additionalInformation.isEmpty()) {
+                builder.additionalInformation(additionalInformation);
+            }
+
+            idmClient.setOIDCClient(tenant, builder.build());
+        } catch (Exception e) {
+            logger.error("Unable to update metadata for client: " + resource.getClientId(), e);
+            throw new VmidentityDataAccessException("Unable to update metadata for client: " + resource.getClientId());
+        }
+
+        return retrieve(resource.getClientId());
+    }
+
+    public static ClientMetadata mapOIDCClient(OIDCClient client, String tenant) {
+        ClientMetadata clientMetadata = new ClientMetadata();
+        clientMetadata.setClientId(client.getClientId());
+        clientMetadata.setIdentityZoneId(tenant); // TODO check if this should be tenant or the zone (only matters for default)
```

Comment:
That was a TODO from before I broke it out into a separate call. The callers are already doing the mapping - in this case it should just set the `zoneId` to the contents of `tenant` as it's already doing. Sorry for the confusion.